ea-factories/src/factories/balance.ts

```diff
@@ -0,0 +1,168 @@
+
+  // Do we support processing?
+  const supported = config.isSupported(account.coin, account.chain)
+  if (!supported)
+    return { ...account, warning: WARNING_UNSUPPORTED_PARAMS + ` ${account.chain} ${account.coin}` }
+
+  // If warning, clear and continue to processing
+  const { warning, ...accNoWarning } = account
+  return accNoWarning
+}
+
+const getBalances = async (
+  config: BalanceConfig,
+  accounts: Account[],
+  getBalance: GetBalance,
+): Promise<BalanceResponse[]> =>
+  Promise.all(
+    accounts.map((acc) => {
+      if (acc.warning) return { result: acc }
+      return getBalance(acc, config)
+    }),
+  )
+
+const reduceBalances = (config: BalanceConfig, responses: BalanceResponse[]) =>
+  responses.reduce<SequenceResponseData<Account>>(
+    (accumulator, current) => {
+      if (config.verbose) accumulator.responses = [...accumulator.responses, current]
+      accumulator.result = [...accumulator.result, current.result]
+      return accumulator
+    },
+    { responses: [], result: [] },
+  )
+
+export const getBalancesBatch = async (
+  config: BalanceConfig,
+  accGroups: Account[][],
+  getBalances: GetBalances,
+): Promise<BalancesResponse[]> =>
+  Promise.all(accGroups.map((accounts) => getBalances(accounts, config)))
+
+const reduceBalancesBatch = (config: BalanceConfig, responses: BalancesResponse[]) =>
+  responses.reduce<SequenceResponseData<Account>>(
+    (accumulator, current) => {
+      if (config.verbose) accumulator.responses = [...accumulator.responses, current]
+      accumulator.result = [...accumulator.result, ...current.result]
+      return accumulator
+    },
+    { responses: [], result: [] },
+  )
+
+const inputParams = {
+  dataPath: false,
+  confirmations: false,
+}
+
+export const make: ExecuteFactory<BalanceConfig> = (config) => async (input) => {
+  const validator = new Validator(input, inputParams)
+  if (validator.error) throw validator.error
+  if (!config) throw new Error('No configuration supplied')
+  if (!config.getBalance && !config.getBalances)
+    throw new Error('Request handling logic not supplied')
+  config.confirmations = validator.validated.confirmations || DEFAULT_CONFIRMATIONS
+  const jobRunID = validator.validated.id
+  const data = validator.validated.data
+  const dataPath = validator.validated.dataPath || DEFAULT_DATA_PATH
+  const accounts = requireArray(jobRunID, data, dataPath).map((acc) =>
+    toValidAccount(jobRunID, acc, config),
+  )
+
+  let response
+
+  if (config.getBalances) {
+    const grouped = groupBy(accounts, (acc) => `${acc.coin}-${acc.chain}`)
+    const responses = await getBalancesBatch(
+      config,
+      Array.from(grouped.values()),
+      config.getBalances,
+    )
+    response = reduceBalancesBatch(config, responses)
+  }
+  if (config.getBalance) {
+    const responses = await getBalances(config, accounts, config.getBalance)
+    response = reduceBalances(config, responses)
+  }
```

Comment:
We need just one workflow here, not two different ones that are almost the same.

First, define the function you require:

```javascript
const getBalances = config.getBalances || // create it from config.getBalance
```

Then group the data and execute M requests for M groups:

```javascript
const key = (acc) => `${acc.coin}-${acc.chain}`
const requests = groupBy(accounts, key)
  .values()
  .map(group => getBalances(group, config))
const responses = await Promise.all(requests)
```

Then reduce:

```javascript
const data = {
  responses: responses.map(r => r.payload),
  result: responses.flatMap(r => r.result),
}
```

Then in the end act on the `verbose` config:

```javascript
if (!config.verbose) delete data.responses;
```

Reply:
This is a good suggestion. Ended up being a lot cleaner. 

I originally split these because I was having type issues, but this way is a good approach.